src/App.php

```diff
@@ -434,29 +450,49 @@ public function add($seed, $region = null)
      * Runs app and echo rendered template.
      *
      * @throws \atk4\core\Exception
+     * @throws ExitApplicationException
      */
     public function run()
     {
-        $this->run_called = true;
-        $this->hook('beforeRender');
-        $this->is_rendering = true;
+        $is_exit_exception = false;
 
-        // if no App layout set
-        if (!isset($this->html)) {
-            throw new Exception(['App layout should be set.']);
-        }
+        try {
+            ob_start();
+            $this->run_called = true;
+            $this->hook('beforeRender');
+            $this->is_rendering = true;
 
-        $this->html->template->set('title', $this->title);
-        $this->html->renderAll();
-        $this->html->template->appendHTML('HEAD', $this->html->getJS());
-        $this->is_rendering = false;
-        $this->hook('beforeOutput');
+            // if no App layout set
+            if (!isset($this->html)) {
+                throw new Exception(['App layout should be set.']);
+            }
+
+            $this->html->template->set('title', $this->title);
+            $this->html->renderAll();
+            $this->html->template->appendHTML('HEAD', $this->html->getJS());
+            $this->is_rendering = false;
+            $this->hook('beforeOutput');
 
-        if (isset($_GET['__atk_callback']) && $this->catch_runaway_callbacks) {
-            $this->terminate('!! Callback requested, but never reached. You may be missing some arguments in '.$_SERVER['REQUEST_URI']);
+            if (isset($_GET['__atk_callback']) && $this->catch_runaway_callbacks) {
+                $this->terminate(
+                    '!! Callback requested, but never reached. You may be missing some arguments in '.$_SERVER['REQUEST_URI']
+                );
+            }
+            echo $this->html->template->render();
+        } catch (ExitApplicationException $e) {
+            $is_exit_exception = true;
```

Comment:
wrap all App->run i a try catch that catch only exit application exception and call exit app if raised